migration.js

```diff
@@ -0,0 +1,49 @@
+const path = require('path');
+const pull = require('pull-stream');
+const FlumeLog = require('flumelog-offset');
+const AsyncFlumeLog = require('async-flumelog');
+const bipf = require('bipf');
+const jsonCodec = require('flumecodec/json');
+const debug = require('debug')('ssb:db2:migration');
+
+const blockSize = 64 * 1024;
+
+exports.init = function (_sbot, config) {
+  const oldLogPath = path.join(config.path, 'flume', 'log.offset');
+  const newLogPath = path.join(config.path, 'flumelog.bipf');
+
+  const oldLog = FlumeLog(oldLogPath, {blockSize, codec: jsonCodec});
+  const newLog = AsyncFlumeLog(newLogPath, {blockSize});
+
+  let dataTransferred = 0;
+
+  pull(
+    oldLog.stream({seqs: false, codec: jsonCodec}),
+    // oldLog.stream({seqs: false, live: true, codec: jsonCodec}), // FIXME: live
+    pull.map(function (data) {
+      const len = bipf.encodingLength(data);
+      const buf = Buffer.alloc(len);
+      bipf.encode(data, buf, 0);
+      return buf;
+    }),
+    function sink(read) {
+      read(null, function next(err, data) {
+        if (err && err !== true) throw err;
+        if (err) {
+          debug('done');
+          if (config._db2migrationCB) newLog.onDrain(() => {
+            newLog.close(config._db2migrationCB)
```

Comment:
Opinion : as the final step move log.offset to log.offset.old

Motivation is to avoid forking feeds... But now I realize that wouldn't work completely. Maybe ssb-db would benefit from a pr which checks if flumelog.bipf before creating a new log.. M

Reply:
Hmm! Very good point about forking! It seems to me that if both logs are in use, we have to have a strong guarantee that only one of them are written.

Reply:
I think the documentation should be very explicit about this. There can only be 1 "master" log, so you have to write to the old log as long as that is still used. Maybe add a guard to add and publish that will detect if there is an active migration and abort.

Reply:
Yes I thought about that same guard idea some minutes ago. We could check if the old log file exists and decline `add` and `publish` on db2 until it's deleted. Why a delete: because other apps or processes may write to the old log (have to assume that the currently running process doesn't have exclusive access to ~/.ssb)

Reply:
But with a delete if another app starts it will just make a new log. Likely no one will replicate it tho

Reply:
There has to be some smart way of forcing just one source of truth, I'm not sure yet which method, but there should be _some method_ is my point

Reply:
I've got it! Leave a log.offset which is not a valid log file at all, is instead a picture of a pony.
Any app trying to load an offset log will instead get a pony which will clog its machinery.

Ugly but works

Reply:
![](https://duckduckgo.com/?q=pony&t=fpas&iax=images&ia=images&iai=http%3A%2F%2Fwww.bouncezoo.com%2Fuploads%2F2%2F5%2F9%2F8%2F25983532%2F6549118_orig.jpg)

Reply:
gonna 'resolve' this because I made a TODO for it in the orig post, and this inline comment refers to disappeared code

Reply:
Ok, as long as we don't lose these ponies, it took me ages to find them